mcp/src/minecraft/com/elementfx/tvp/ad/inventory/ContainerGoblinWorkbench.java

```diff
@@ -0,0 +1,169 @@
+package com.elementfx.tvp.ad.inventory;
+
+import javax.annotation.Nullable;
+import com.elementfx.tvp.ad.item.crafting.GoblinCraftingManager;
+
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.InventoryPlayer;
+import net.minecraft.init.Blocks;
+import net.minecraft.inventory.Container;
+import net.minecraft.inventory.ContainerWorkbench;
+import net.minecraft.inventory.IInventory;
+import net.minecraft.inventory.InventoryCraftResult;
+import net.minecraft.inventory.InventoryCrafting;
+import net.minecraft.inventory.Slot;
+import net.minecraft.inventory.SlotCrafting;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.CraftingManager;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+
+
+public class ContainerGoblinWorkbench extends Container
+{
+	/** The crafting matrix inventory (3x3). */
+    public InventoryCrafting craftMatrix = new InventoryCrafting(this, 3, 3);
+    public IInventory craftResult = new InventoryCraftResult();
+    private final World worldObj;
+
+    /** Position of the workbench */
+    private final BlockPos pos;
+
+    public ContainerGoblinWorkbench(InventoryPlayer playerInventory, World worldIn, BlockPos posIn)
+    {
+        this.worldObj = worldIn;
+        this.pos = posIn;
+        
+        this.addSlotToContainer(new SlotGoblinCrafting(playerInventory.player, this.craftMatrix, this.craftResult, 0, 124, 35));
+
+        for (int i = 0; i < 3; ++i)
+        {
+            for (int j = 0; j < 3; ++j)
+            {
+                this.addSlotToContainer(new Slot(this.craftMatrix, j + i * 3, 30 + j * 18, 17 + i * 18));
+            }
+        }
+
+        for (int k = 0; k < 3; ++k)
+        {
+            for (int i1 = 0; i1 < 9; ++i1)
+            {
+                this.addSlotToContainer(new Slot(playerInventory, i1 + k * 9 + 9, 8 + i1 * 18, 84 + k * 18));
+            }
+        }
+
+        for (int l = 0; l < 9; ++l)
+        {
+            this.addSlotToContainer(new Slot(playerInventory, l, 8 + l * 18, 142));
+        }
+
+        this.onCraftMatrixChanged(this.craftMatrix);
+    }
+
+    /**
+     * Callback for when the crafting matrix is changed.
+     */
+    public void onCraftMatrixChanged(IInventory inventoryIn)
+    {
+        this.craftResult.setInventorySlotContents(0, GoblinCraftingManager.getInstance().findMatchingRecipe(this.craftMatrix, this.worldObj));
```

Comment:
Is this and the result slot type the only differences between all of the ContainerXWorkbench classes? Do you expect we'll need custom classes for other reasons?